src/marked.js

```diff
@@ -107,29 +115,52 @@ export function marked(src, opt, callback) {
 
   function onError(e) {
     e.message += '\nPlease report this to https://github.com/markedjs/marked.';
+
     if (opt.silent) {
-      return '<p>An error occurred:</p><pre>'
+      const msg = '<p>An error occurred:</p><pre>'
         + escape(e.message + '', true)
         + '</pre>';
+      if (opt.async) {
+        return Promise.resolve(msg);
+      }
+      return msg;
+    }
+
+    if (opt.async) {
+      return Promise.reject(e);
     }
     throw e;
   }
 
+  if (opt.async) {
+    return Promise.resolve(opt.hooks ? opt.hooks.preprocess(src) : src)
```

Comment:
it seems like we might need to pass `opt` in to the hooks and pass `opt` out if we want to be able to change the options in preprocess like in the front-matter example. Currently if someone uses `marked.parse(md, { headerIds: false })` the hook that looks at `this.defaults` won't see the `headerIds` option changed because it is not changed in defaults.

Reply:
I added options to the hooks object and passed that as `this` similar to the way we do in tokenizer and renderer